Paper: Python-Based GeoImagery Dataset Development for Deep Learning-Driven Forest Wildfire Detection (#916)
scipy-conference/scipy_proceedings: papers/Valeria_Martin/main.tex

\documentclass{article}
\usepackage{graphicx} % Required for inserting images
\usepackage{listings}
\usepackage{float}
\usepackage{color}
\usepackage{booktabs} 
\definecolor{codegreen}{rgb}{0,0.6,0}
\definecolor{codegray}{rgb}{0.5,0.5,0.5}
\definecolor{codepurple}{rgb}{0.58,0,0.82}
\definecolor{backcolour}{rgb}{0.95,0.95,0.92}
\usepackage[numbers]{natbib}
\usepackage{hyperref}

\lstdefinestyle{mystyle}{
    backgroundcolor=\color{backcolour},   
    commentstyle=\color{codegreen},
    keywordstyle=\color{magenta},
    numberstyle=\tiny\color{codegray},
    stringstyle=\color{codepurple},
    basicstyle=\footnotesize,
    breakatwhitespace=false,         
    breaklines=true,                 
    captionpos=b,                    
    keepspaces=true,                 
    numbers=left,                    
    numbersep=5pt,                  
    showspaces=false,                
    showstringspaces=false,
    showtabs=false,                  
    tabsize=2
}

\lstset{style=mystyle}

\begin{abstract}

> In recent years, integrating satellite imagery with deep learning (DL) architectures has become an effective approach for environmental monitoring tasks, including forest wildfire detection. Nevertheless, this integration requires substantial high-quality labeled data to train the DL models accurately. Leveraging the capabilities of Google Earth Engine’s Python API, this study introduces a streamlined Python-based methodology to efficiently build, label, process, and evaluate a large-scale bi-temporal high-resolution satellite imagery dataset for DL-driven forest wildfire detection. Known as the California Wildfire GeoImaging Dataset (CWGID), this dataset comprises over 100,000 labeled 'before' and 'after' wildfire image pairs. An analysis of the dataset using pre-trained and adapted Convolutional Neural Network (CNN) architectures, such as VGG16 and EfficientNet, achieved accuracies of respectively 82\% and 93\%. The pipeline described in this paper explains how Python can be used to build a high-resolution satellite imagery dataset and how to train DL models with it, providing an efficient framework for improving environmental monitoring efforts. 

Reviewer: A few more specific comments / questions here: 
* I think some of the language used in the abstract makes it a little difficult to understand what specifically you did and what the impact was (see the next two bullets for specific examples).  This seems like a great start, but I think adjusting the language a bit and being really clear about what the work entailed and the impact is would help the readers understand your work.
* I wonder if rather than saying "integrating satellite imagery with DL architectures" something like "using satellite imagery with DL architectures" or "leveraging satellite imagery with DL architectures" might better convey what you're doing in this paper.  
* Similarly, I wonder if using something like "process" rather than "build" in this section, i.e. "explains how Python can be used to build a high-resolution satellite imagery dataset" might better convey what you're doing. Or if "process" is insufficient, perhaps describing that the dataset was created from pre-existing imagery and something about the what the dataset is / does.

Author: Thank you for your constructive feedback. I  have replaced "integrating satellite imagery with DL architectures" with "using satellite imagery with DL architectures" to more accurately convey the approach taken in this study.
I have changed "explains how Python can be used to build a high-resolution satellite imagery dataset" to "explains how Python can be used to gather and process high-resolution satellite imagery datasets"

\end{abstract}

\section{Introduction}\label{introduction}

> Generally, building a satellite imagery dataset involves a time-intensive approach. Recently, Google Earth Engine (GEE) \citep{gorelick2017google} has revolutionized this process by providing an extensive, cloud-based platform for the efficient collection, processing, and analysis of satellite imagery. Furthermore, GEE’s Python API allows its users to easily query their platform and automatically download cloud-free large-scale satellite imagery datasets from multiple satellite collections, such as Sentinel-2. This technology has facilitated environmental monitoring by providing automatic and timely access to high-quality satellite imagery. 

Reviewer: "Generally, building a satellite imagery dataset involves a time-intensive approach."
* Adding some specifics to this could help readers understand what pain points GEE might address.

"by providing automatic and timely access"
* I'm not quite sure what is meant by "automatic" here.  "programmatic", perhaps?
* "timely" leads me to think that there's low latency with respect to availability of the data.  Is this data processed in real time and indeed available with low latency?  or maybe you're referring to something else?

Author: 1.	 I appreciate your suggestions for improving clarity. I added the specifics you suggested.  I used "programmatically" to describe how GEE allows users to automate the process of querying and downloading satellite imagery through coding, providing a more accurate description. I changed "timely access" to "easily accessible" to avoid implying real-time processing and to emphasize the ease with which users can access pre-processed high-quality satellite imagery through GEE.

> Satellite imagery-based change detection and forest monitoring have traditionally depended on manually identifying specific features and applying predefined algorithms and models, including differential analysis, thresholding techniques, and clustering and classification algorithms. However, these algorithms and models often fail to capture the full complexity of the studied data. Thus, integrating deep learning (DL) methods with satellite imagery offers a more dynamic and precise approach, one capable of handling the patterns and variability associated with imagery data. 

Reviewer: "Thus, integrating deep learning (DL) methods with satellite imagery offers a more dynamic and precise approach, one capable of handling the patterns and variability associated with imagery data"
* You speak to the limitations of traditional techniques before this statement, but not why DL methods might offer improvements.  This could use some references and/or rewording probably.

Author: 2.	I have revised the paragraph to more clearly articulate why deep learning (DL) methods offer improvements over traditional techniques for satellite imagery-based change detection and forest monitoring. The updated paragraph now includes specific reasons and references to support these claims.

> Python facilitates the use of DL in environmental monitoring by providing a rich ecosystem of libraries and tools, such as TensorFlow \citep{tensorflow2015-whitepaper}, which contains multiple existing DL architectures that can be adapted and used with satellite imagery. Nevertheless, the integration of DL and Remote Sensing images requires multiple processing steps, such as having smaller imagery tiles and adapting the models to use GeoTIFF data, among others. 

Reviewer: "Nevertheless, the integration of DL and Remote Sensing images requires multiple processing steps, such as having smaller imagery tiles and adapting the models to use GeoTIFF data, among others"
* Spelling out these steps here could be helpful to guide the readers and let them know what might be coming.

Author: I have changed this paragraph to spell out the necessary steps for integrating deep learning (DL) with remotely sensed images. This will provide readers with a clearer understanding of the process involved.

# Freeze the convolutional base to prevent weights from being updated
for layer in base_model.layers:
    layer.trainable = False

# Add custom layers on top of the base model
x = Flatten()(base_model.output) # Flatten the output to make it suitable for dense layers
x = Dense(1024, activation='relu')(x) # Add a dense layer with 1024 units and ReLU activation
output = Dense(1, activation='sigmoid')(x) # Output layer with a sigmoid activation for binary classification

# Construct the complete model
model = Model(inputs=base_model.input, outputs=output)

# Compile the model with Adam optimizer and binary cross-entropy loss
model.compile(optimizer=Adam(learning_rate=0.0001),
              loss='binary_crossentropy',
              metrics=['accuracy', Precision(), Recall()])

\end{lstlisting}

The VGG16 model is trained and validated on 95594 three-channel RGB images (45\% of the data, divided as follows: 80\% for training and
10\% for validation) and is tested on 10621 images (5\% of the data). Training took 1953 minutes for 10 epochs. 
\subsection{EfficientNet Implementation}
\begin{figure}[H]
    \centering
    \includegraphics[width=0.9\textwidth]{6channel.png}

>     \caption{Representation of a 6 Channel RGB GeoTIFF Input. A: Representation of a 3-channel RGB GeoTIFF forested area \textit{before} a wildfire B: Visual example of a 3-channel RGB GeoTIFF forested area \textit{after} a wildfire.}. \label{fig1}

Reviewer 2: ```suggestion
    \caption{Representation of a 6 Channel RGB GeoTIFF Input. A: Representation of a 3-channel RGB GeoTIFF forested area \textit{before} a wildfire B: Visual example of a 3-channel RGB GeoTIFF forested area \textit{after} a wildfire.} \label{fig1}
```

On top of that, the dataset's bi-temporal collection of pre- and post-wildfire satellite imagery also offers the possibility to use different and accurate approaches for detecting forest wildfires with satellite imagery. The comparative analysis of VGG16 and EfficientNet shows the advantage of employing a bitemporal approach to satellite imagery analysis under a single input. By evaluating \textit{before} and \textit{after} images as coupled inputs, the model identifies changes that a single post-event imagery might not reveal, as shown by VGG16. 

\section{Conclusion}

> This paper illustrates a Python methodology that integrates DL with satellite imagery to enhance environmental monitoring. Specifically, the development of the California Wildfire GeoImaging Dataset (CWGID) and its evaluation using DL architectures are outlined. The approach explained throughout this work improves the accurate and timely detection of forest wildfires and can be applied to a broad range of environmental issues, beyond just wildfire detection.

Reviewer: "The approach explained throughout this work improves the accurate and timely detection of forest wildfires"
* Can you speak more to why and how your work does this?

Author: Thank you! I have removed the sentence and emphasized how our approach enables accurate detection.

\begin{enumerate}
    \item Data Acquisition: Collect satellite imagery from sources such as Google Earth Engine or 		 other satellite image providers.
    \item Preprocessing: Convert satellite images into formats suitable for analysis.
    \item Image Tiling: Divide large satellite images into smaller tiles to fit the input image size used in DL models.
    \item Data Annotation: Label the images to create a ground truth dataset for training DL models.
    \item Data Augmentation: Applying transformations such as rotations and flips to increase the diversity of the training dataset.
\end{enumerate}
Python’s tools help manage these steps. For example, libraries like rasterio, Tifffile and GeoPandas, can be used to process and transform satellite imagery data into formats suitable for DL models. This reduces the manual effort required and ensures that the data pipeline is efficient.

> Furthermore, DL algorithms require a substantial amount of labeled data to effectively learn and identify change \citep{Alzubaidi2021ReviewOD}. Therefore, the development of labeled high-resolution satellite imagery datasets is important and relevant for addressing environmental problems. Currently, the availability of labeled high-quality satellite imagery datasets is an obstacle to developing DL models for environmental change detection \citep{Adegun2023}.  Furthermore, manually labeling data can be time-consuming, but with Python’s tools and leveraging historic and reliable shapefile data, this process can be streamlined.

Reviewer: ```suggestion
Furthermore, DL algorithms require a substantial amount of labeled data to effectively learn and identify change \citep{Alzubaidi2021ReviewOD}. Therefore, the development of labeled high-resolution satellite imagery datasets is important and relevant for addressing environmental problems. Currently, the availability of labeled high-quality satellite imagery datasets is an obstacle to developing DL models for environmental change detection \citep{Adegun2023}.  Manually labeling data can be time-consuming, but with Python’s tools and leveraging high quality historical data stored in shapefiles, this process can be streamlined.

```

Rewording to avoid repetition and change from "historic" to "historical".

This paper presents a methodology, entirely implemented in Python, to streamline the creation and evaluation, via DL, of satellite imagery datasets. The methodology covers the entire workflow: from data acquisition, labelling, and preprocessing to model adaptation, training, and evaluation. Specifically, this approach is applied to gather and validate a high-resolution dataset for forest wildfire detection, the California Wildfire GeoImaging Dataset (CWGID). Additionally, this methodology can be adapted for various environmental monitoring tasks, showing its versatility in studying and responding to different environmental changes.

\section{Building a Sentinel-2 Satellite Imagery Dataset}
To construct the CWGID, a multi-step process is needed. 
\subsection{Gathering and Refining Historic Wildfire Polygon Data from California}

> The initial step is to gather geo-referenced forest wildfire polygon data from California, sourced from the Fire and Resource Assessment Program (FRAP) maintained by the California Department of Forestry and Fire Protection \citep{california_department_of_forestry_and_fire_protection_2024}. This FRAP data includes perimeters of past wildfires and serves as the geographic reference needed to select satellite imagery with GEE. Figure \ref{fig2} illustrates the polygons from the FRAP. The polygons delineated in purple represent areas affected by wildfires in forested regions. These delineated polygons are used to create the CWGID. 

Reviewer: ```suggestion
The initial step is to gather georeferenced forest wildfire polygon data from California, sourced from the Fire and Resource Assessment Program (FRAP) maintained by the California Department of Forestry and Fire Protection \citep{california_department_of_forestry_and_fire_protection_2024}. This FRAP data includes perimeters of past wildfires and serves as the geographic reference needed to select satellite imagery with GEE. Figure \ref{fig2} illustrates the polygons from the FRAP. The polygons delineated in purple represent areas affected by wildfires in forested regions. These delineated polygons are used to create the CWGID. 

```

To address this imbalance and enhance the model's accuracy in detecting wildfire-affected areas, data augmentation techniques are implemented. Specifically, functions using the rasterio and NumPy \citep{numpy} libraries are developed to perform image transformations, which include rotating the GeoTIFF tiles by 90°, 180°, and 270°, and flipping them horizontally and vertically. These transformations are applied to the imagery and the mask tiles that contain fire polygons and to their corresponding pre-wildfire RGB GeoTIFF tiles.

Finally, the augmentation process increased the diversity of the training data and resulted in a total of 111,093 pairs of labeled RGB GeoTIFF image tiles. The final dataset consists of 82,011 negative instances (no wildfire damage) and 29,082 positive instances (wildfire damage), significantly improving the class balance and reducing the risk of overfitting.

> Figure X illustrates the data augmentation process applied to the satellite imagery and its labels. The original image tile is shown alongside its augmented versions, which include rotations and flips. This augmentation increases the diversity of the dataset, helping to balance the classes and improve the model's ability to detect wildfire-affected areas accurately.

Reviewer: Did you mean to reference a specific figure here?  Looks like there's still a placeholder "Figure X"

Author: I did! Thank you. I fixed this paragraph.

\begin{figure}[H]
    \centering
    \includegraphics{augment.png}
    \caption{Representation of the data augmentation performed to the 256px*256px tiles to increase positive examples of wildfire damage. \label{fig0}}
\end{figure}

\section{Evaluating the CWGID Using Convolutional Neural Networks}
The CWGID is tested for demonstration purposes using the well-known CNN architectures VGG16 and EfficientNet, both implemented using Python's TensorFlow’s Keras API.

\subsection{VGG16 Implementation}
VGG16 \citep{Simonyan15} is a deep CNN designed for image processing and classification tasks. It consists of 13 convolutional layers, 5 pooling layers, and 3 fully connected layers. This model is adapted to train on the dataset and detect positive and negative instances of forest wildfires. The VGG16 architecture only trains on the post-wildfire 3-channel RGB imagery, which contains both.

> To use this CNN architecture with GeoTIFF satellite images, which typically store geographic data not supported by most deep-learning libraries, a custom function is implemented to process and feed the data to the model using the rasterio library and the shuffle function from the Scikit-learn library \citep{sklearn1,sklearn2}. This function (see Code \ref{lst}) is crucial for handling the unique structure of GeoTIFF files and converting them into a format suitable for deep learning. It is designed to read batches of GeoTIFF files, shuffle them to ensure randomness, and process them into a three-dimensional array compatible with VGG16. Here is how it works:

Reviewer: It looks like your code reference here is broken.

Author: You are right, thank you. I believe it is now fixed.

\begin{itemize}
\item File Paths Initialization: The function starts by taking the file paths of the GeoTIFF images and shuffling them using the shuffle function from Scikit-learn to ensure the batches are randomly ordered.

\item Batch Processing: It processes the images in batches of 32 files. For each batch, the function initializes empty lists for images and labels.

\item Reading and Transforming Images: It reads each GeoTIFF file using rasterio. The images are then transformed to have the channels last (i.e., changing the format from \textbf{(channels, height, width)} to \textbf{(height, width, channels))} to use with Keras.

> \item Labels are assigned based on the presence of "Damaged" in the file paths. Specifically, if the file path contains the string "/Damaged/", the image is labeled as 1 (indicating it shows a damaged area). If the file path does not contain this string, the image is labeled as 0 (indicating it shows an undamaged area). In this method, the ground-truth masks are not directly used but are leveraged to define the imagery file paths. Hence, if the mask of an image had at least one pixel with a value of one, the image was redirected to the "Damaged" files.

Reviewer: I wonder if this threshold (i.e. how many pixels or what percentage of pixels should be used to classify an image as "damaged") might be something you could experiment with in the future (perhaps you've already thought about this).  

Using a threshold of 1 pixel seems like it could potentially lead to false positives and other issues.

Author: Thank you for your feedback. It is a good catch. I understand the concern regarding the threshold used to classify an image as "damaged." The one-pixel threshold was chosen based on the presence of a polygon indicating wildfire presence, which makes it less likely for these pixels to be isolated. We checked the masks and less than 10% of them have less than 1000 pixels with one in it.   However, I agree that this choice could potentially lead to false positives. I can explore different thresholds, such as a specific percentage of damaged pixels within an image, to improve classification accuracy in future experiments.

Reviewer: To be clear this was more of a suggestion to think about (sounds like you already have to some degree 😄 ) or maybe for future work.

> \item Yielding Batches: Finally, the function yields batches of images and labels as numpy arrays, which are then fed into the VGG16 model for training.

Reviewer: ```suggestion
\item Yielding Batches: Finally, the function yields batches of images and labels as NumPy arrays, which are then fed into the VGG16 model for training.
```

Author: Done!

\end{itemize}

> To use this CNN architecture with GeoTIFF satellite images, which typically store geographic data not supported by most deep-learning libraries, a custom function to feed the data to the model is built using the rasterio library and the shuffle function from the Scikit-learn library \citep{sklearn1,sklearn2} (see Code \ref{lst:custom_generator}). 

Reviewer: Similarly, the code reference here seems to be broken.

    11/20 & 0.9537 & 0.9590 & 0.8655 & 0.9500 & 0.9333 & 0.9386 & 0.7959 & 0.8672 \\
    12/20 & 0.9573 & 0.9644 & 0.8739 & 0.9576 & 0.9363 & 0.9516 & 0.7959 & 0.8650 \\
    13/20 & 0.9588 & 0.9665 & 0.8776 & 0.9622 & 0.9333 & 0.9606 & 0.7757 & 0.8712 \\
    \bottomrule
    \end{tabular}
\end{table}

There is a slight decrease in the validation performance metrics when compared to the training performance metrics. However, the validation metrics remain high, which suggests that the model is generalizing well to unseen data but the consistent gap between training and validation performance metrics indicates a slight degree of overfitting for the EfficientNetB0 application. 

\subsection{DL Results}
The results in Table \ref{tab:2} compare the performance of the two neural network architectures used on the CWGID. Each architecture's performance metrics are assessed, including loss, accuracy, precision, recall, and training time. Also, the percentage of the CWGID used to train each architecture is specified. 

Both models are trained with subsets of the CWGID dataset, with the 6-channel input EF EfficientNet model using 25\% of the data and the VGG16 model using 50\% of the data. 

> The 6-channel input EfficientNet achieved the lowest loss (0.178) and highest accuracy (92.6\%), making it the most effective model in terms of overall performance. Its precision was at 92.5\% and its recall was 81.9\%, indicating a strong ability to identify damaged areas correctly with some missed detections. VGG16, despite processing a larger dataset, had a higher loss (1.220) but maintained an accuracy of 83.2\%. Its precision was 74.9\% and its recall was 58.8\%. This architecture took a longer training time. The EfficientNet model shows a relatively faster training time when considering the epochs and the amount of data used.  The 6-channel input EfficientNet demonstrated superior performance in accurately classifying areas affected by wildfires with high precision and accuracy, making it an efficient choice for applications where quick and accurate assessments are critical. Moreover, it is important to mention that VGG16 was tested with Six-Channel inputs but proved computationally intensive.

Reviewer: ```suggestion
The 6-channel input EfficientNet achieved the lowest loss (0.178) and highest accuracy (92.6\%), making it the most effective model in terms of overall performance. Its precision was at 92.5\% and its recall was 81.9\%, indicating a strong ability to identify damaged areas correctly with some missed detections. VGG16, despite processing a larger dataset, had a higher loss (1.220) but maintained an accuracy of 83.2\%. Its precision was 74.9\% and its recall was 58.8\%. This architecture took a longer training time. The EfficientNet model shows a relatively faster training time when considering the epochs and the amount of data used.  The 6-channel input EfficientNet demonstrated superior performance in accurately classifying areas affected by wildfires with high precision and accuracy, making it an efficient choice for applications where quick and accurate assessments are critical. Moreover, it is important to mention that VGG16 was tested with six-channel inputs but proved computationally intensive.
```

Author: Fixed!
